File: CSC310-Python/Module13-Shortest-Paths/debug03/floyd_warshall.py
"""Floyd-Warshall all-pairs shortest paths (CLRS Section 23.2).

floyd_warshall(W) takes the weight matrix of a directed graph with no
negative-weight cycles and returns the matrix D of shortest-path weights
together with a predecessor matrix for reconstructing paths.  The
dynamic-programming idea: d[i][j] is improved by allowing one more vertex
k to serve as an intermediate vertex on the path from i to j.  The main
block runs the algorithm on the five-vertex example graph of CLRS Chapter 23,
prints both
matrices, and reconstructs a few shortest paths.
"""

INF = float("inf")


def floyd_warshall(W):
    n = len(W)
    d = [row[:] for row in W]
    pred = [[None if i == j or W[i][j] == INF else i for j in range(n)]
            for i in range(n)]
    for i in range(n):
        for j in range(n):
            for k in range(n):
                if d[i][k] + d[k][j] < d[i][j]:
                    d[i][j] = d[i][k] + d[k][j]
                    pred[i][j] = pred[k][j]
    return d, pred


def path_between(pred, i, j):
    path = [j]
    while j != i and len(path) <= len(pred):
        j = pred[i][j]
        path.append(j)
    return " -> ".join(str(v + 1) for v in reversed(path))


def print_matrix(title, M):
    print(title)
    for row in M:
        print("  " + " ".join(f"{'inf' if x == INF else x:>4}" for x in row))


def main():
    # The five-vertex example graph of CLRS Chapter 23, vertices 1..5 stored as 0..4
    W = [
        [0,   3,   8,   INF, -4],
        [INF, 0,   INF, 1,   7],
        [INF, 4,   0,   INF, INF],
        [2,   INF, -5,  0,   INF],
        [INF, INF, INF, 6,   0],
    ]
    print_matrix("weight matrix W:", W)
    D, pred = floyd_warshall(W)
    print_matrix("shortest-path weights D:", D)
    print("sample paths:")
    for i, j in [(0, 1), (0, 2), (4, 0)]:
        print(f"  {i + 1} to {j + 1}: {path_between(pred, i, j)}   (weight {D[i][j]})")


if __name__ == "__main__":
    main()

File: CSC310-Cpp/_solutions/Module13-Shortest-Paths/debug03/floyd_warshall.cpp
// floyd_warshall.cpp
// All-pairs shortest paths with the Floyd-Warshall algorithm (CLRS 23.2).
// d[i][j] is the shortest-path weight from i to j using only intermediate
// vertices from an ever-growing prefix of the vertex set.
#include <iomanip>
#include <iostream>
#include <vector>

const int INF = 1000000000;

void printMatrix(const std::vector<std::vector<int>>& d) {
    for (const auto& row : d) {
        for (int x : row) {
            if (x >= INF) std::cout << std::setw(5) << "inf";
            else std::cout << std::setw(5) << x;
        }
        std::cout << '\n';
    }
}

std::vector<std::vector<int>> floydWarshall(std::vector<std::vector<int>> d) {
    const int n = static_cast<int>(d.size());
    for (int k = 0; k < n; ++k) {
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (d[i][k] < INF && d[k][j] < INF && d[i][k] + d[k][j] < d[i][j]) {
                    d[i][j] = d[i][k] + d[k][j];
                }
            }
        }
    }
    return d;
}

int main() {
    // CLRS's Floyd-Warshall example (§23.1), vertices numbered 0..4 (book's 1..5)
    const int n = 5;
    std::vector<std::vector<int>> w(n, std::vector<int>(n, INF));
    for (int i = 0; i < n; ++i) w[i][i] = 0;
    w[0][1] = 3;  w[0][2] = 8; w[0][4] = -4;
    w[1][3] = 1;  w[1][4] = 7;
    w[2][1] = 4;
    w[3][0] = 2;  w[3][2] = -5;
    w[4][3] = 6;

    std::cout << "weight matrix W:\n";
    printMatrix(w);
    std::vector<std::vector<int>> d = floydWarshall(w);
    std::cout << "shortest-path weights D:\n";
    printMatrix(d);
    return 0;
}
